<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.302">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2023-04-04">

<title>HCS7003 - Command-line genomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">HCS7003</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../modules/presentation.html" rel="" target="" aria-current="page">
 <span class="menu-text">Presentation</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    <div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="presentation.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Command-line genomics</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Jelmer Poelstra </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            MCIC Wooster, OSU
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 4, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="what-ill-talk-about-today" class="level3">
<h3 class="anchored" data-anchor-id="what-ill-talk-about-today">What I’ll talk about today</h3>
<p>Today, I will give an overview of the <strong>typical computational environment infrastructure used for genomics projects</strong>.</p>
<p>Most of this can be summarized as “<strong>command-line genomics</strong>”, and I will explain what this entails, and why you need the command line.</p>
<p><br></p>
<p>Of course, I won’t have time to teach you the different components, but I would like to orient you on this topic so that it’s not as much of a black box.</p>
<p>This will hopefully give you a starting point for learning more – and I will point you to some specific resources as well.</p>
<hr>
</section>
<section id="what-kind-of-research-is-this-relevant-for" class="level3">
<h3 class="anchored" data-anchor-id="what-kind-of-research-is-this-relevant-for">What kind of research is this relevant for?</h3>
<div class="incremental">
<ul class="incremental">
<li><p>Any project in which you generate <em>high-throughput sequencing data</em>, e.g.:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Whole-genome sequencing</strong> – de novo assembly, pangenomics, “resequencing”</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Reduced-representation sequencing</strong> (GBS, etc) for population genomics</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Microbiomics</strong> – both shotgun metagenomics and amplicon metabarcoding</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Transcriptomics</strong> with RNAseq</p></li>
</ul>
<p><br></p>
<ul class="incremental">
<li>Other genomics projects like <strong>comparative genomics</strong> with publicly available genomes.</li>
</ul>
<p><br></p>
<ul class="incremental">
<li>But only to some extent to <em>proteomics</em> and <em>metabolomics</em>.</li>
</ul>
</div>
<hr>
</section>
<section id="overview" class="level3">
<h3 class="anchored" data-anchor-id="overview">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p><strong>Command-line genomics?</strong></p></li>
<li><p>Ohio Supercomputer Center (OSC) overview</p></li>
<li><p>Command-line software</p></li>
<li><p>The VS Code editor and the whole game</p></li>
</ol>
<hr>
</section>
<section id="a-typical-computational-infrastructure-for-genomics" class="level3">
<h3 class="anchored" data-anchor-id="a-typical-computational-infrastructure-for-genomics">A typical computational infrastructure for genomics</h3>
<div class="incremental">
<ul class="incremental">
<li><p>A <strong>supercomputer</strong> – in our case, the Ohio Supercomputer Center (OSC)<br>
<em>(Cloud computing is an alternative, but won’t be covered today)</em></p></li>
<li><p>The <strong>Unix shell</strong> (terminal)</p></li>
<li><p>A <strong>text editor</strong> – I recommend and will demonstrate <em>VS Code</em></p></li>
</ul>
</div>
<hr>
</section>
<section id="a-typical-computational-infrastructure-for-genomics-1" class="level3">
<h3 class="anchored" data-anchor-id="a-typical-computational-infrastructure-for-genomics-1">A typical computational infrastructure for genomics</h3>
<ul>
<li><p>A <strong>supercomputer</strong> – in our case, the Ohio Supercomputer Center (OSC)<br>
<em>(Cloud computing is an alternative, but won’t be covered today)</em></p></li>
<li><p>The <strong>Unix shell</strong> (terminal)</p></li>
<li><p>A <strong>text editor</strong> – I recommend and will demonstrate <em>VS Code</em></p></li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Command-line genomics
</div>
</div>
<div class="callout-body-container callout-body">
<p>For the purposes of this talk, I will refer to working with the above elements by running command-line programs as “batch jobs” (non-interactively) as <strong>command-line genomics</strong>.</p>
</div>
</div>
<hr>
</section>
<section id="a-typical-computational-infrastructure-for-genomics-2" class="level3">
<h3 class="anchored" data-anchor-id="a-typical-computational-infrastructure-for-genomics-2">A typical computational infrastructure for genomics</h3>
<ul>
<li><p>A <strong>supercomputer</strong> – in our case, the Ohio Supercomputer Center (OSC)<br>
<em>(Cloud computing is an alternative, but won’t be covered today)</em></p></li>
<li><p>The <strong>Unix shell</strong> (AKA the terminal)</p></li>
<li><p>A <strong>text editor</strong> – I recommend and will demonstrate <em>VS Code</em></p></li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Command-line genomics
</div>
</div>
<div class="callout-body-container callout-body">
<p>For the purposes of this talk, I will refer to working with the above elements by running command-line programs as “batch jobs” (non-interactively) as <strong>command-line genomics</strong>.</p>
</div>
</div>
<ul>
<li><strong>R</strong> or perhaps <em>Python</em> for interactive statistical analysis and visualization.<br>
<em>(I won’t talk more about this other than showing you how you can use R at OSC)</em></li>
</ul>
<hr>
</section>
<section id="what-is-a-supercomputer" class="level3">
<h3 class="anchored" data-anchor-id="what-is-a-supercomputer">What is a supercomputer?</h3>
<p>A highly interconnected set of many computer processors and storage units.<br>
You can think of it simply as <strong>a network of computers</strong>.</p>
<p>Supercomputers are also commonly referred to as <em>High-Performance Computing (HPC) clusters</em> or simply <em>compute clusters</em>.</p>
<hr>
</section>
<section id="why-do-i-need-a-supercomputer" class="level3">
<h3 class="anchored" data-anchor-id="why-do-i-need-a-supercomputer">Why do I need a supercomputer?</h3>
<ul>
<li><p>Your genomics <strong>dataset is often too large</strong> to be handled efficiently, or even at all, by a laptop or desktop computer.</p></li>
<li><p>To <strong>speed up long-running analyses</strong> by using more computing power.</p></li>
<li><p>To <strong>speed up repeated analyses</strong>, like the independent mapping of reads for different samples to a reference genome: these can be run in parallel on a supercomputer.</p></li>
<li><p>It’s also a great place to <strong>store large amounts of data</strong></p></li>
</ul>
<hr>
</section>
<section id="what-is-the-unix-shell" class="level3">
<h3 class="anchored" data-anchor-id="what-is-the-unix-shell">What is the Unix shell?</h3>
<p>A computer’s <strong>shell</strong> is also referred to as a <em>Terminal</em> or “<em>the command line</em>”, and allows you to interact with your computer by <strong>typing commands</strong> rather than pointing-and-clicking.</p>
<p><br></p>
<p>The <em>Unix</em> shell is the shell of Unix-based computers, which include Mac and Linux (but not Windows) operating systems.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<p><br></p>
<p align="center">
<img src="img/shell.png" width="70%">
</p>
<hr>
</section>
<section id="why-do-i-need-to-work-in-the-unix-shell" class="level3">
<h3 class="anchored" data-anchor-id="why-do-i-need-to-work-in-the-unix-shell">Why do I need to work in the Unix shell?</h3>
<div class="incremental">
<ul class="incremental">
<li><p>A genomics project usually involves sequentially running a whole array of bioinformatics programs (or “tools”).</p>
<p>For instance, an RNAseq project may include:<br>
raw read QC =&gt; raw read trimming =&gt; trimmed read mapping =&gt; gene counting</p></li>
</ul>
<p><br></p>
<ul class="incremental">
<li><p><strong>Many of these tools can only be run through a “command-line interface” (CLI)</strong></p>
<p>Even those that have a “graphical user interface” (GUI) are more efficiently and reproducibly run through a CLI:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Efficiency</strong> — A CLI allows you to write a simple loop to run it in the same way for many samples. (In combination with the computing power of a supercomputer, this in turn allows you to process those hundreds of samples <em>in parallel</em>.)</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Reproducibility</strong> — You can easily save all commands and scripts which would allow you to rerun a project rather straightforwardly.</p></li>
</ul>
</div>
<hr>
</section>
<section id="howwhen-can-i-avoid-all-of-this" class="level3">
<h3 class="anchored" data-anchor-id="howwhen-can-i-avoid-all-of-this">How/when can I <em>avoid</em> all of this?</h3>
<p>If you will often do genomics projects like the ones mentioned above, it’s hard to avoid command line genomics as described.</p>
<p>But here are some conditions in which you might reasonably avoid it:</p>
<ul>
<li><p>You’re doing a single genomics project, your main research focus is elsewhere</p></li>
<li><p>You have data which can be analyzed with no or a relatively small command-line-based part, such as proteomics/metabolomics/metabarcoding/RNAseq.</p></li>
</ul>
<p>In such cases, you might be able to get someone else to do the command line work, or you could try <strong><em>Galaxy</em></strong>, a cloud-based bioinformatics platform with a web browser interface and no coding.</p>
<hr>
</section>
<section id="overview-1" class="level3">
<h3 class="anchored" data-anchor-id="overview-1">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p>Command-line genomics?</p></li>
<li><p><strong>Ohio Supercomputer Center (OSC) overview</strong></p></li>
<li><p>Command-line software</p></li>
<li><p>The VS Code editor and the whole game</p></li>
</ol>
<hr>
</section>
<section id="the-ohio-supercomputer-center-osc" class="level3">
<h3 class="anchored" data-anchor-id="the-ohio-supercomputer-center-osc">The Ohio Supercomputer Center (OSC)</h3>
<div class="incremental">
<ul class="incremental">
<li><p>The <strong>Ohio Supercomputer Center (OSC)</strong> provides computing resources to researchers (and others) across Ohio.</p>
<p>OSC has two individual supercomputers/clusters (named Owens and Pitzer), and lots of infrastructure for their usage.</p></li>
<li><p>Research usage is charged but at <a href="https://www.osc.edu/content/academic_fee_model_faq">heavily subsidized rates</a>, and most or all OSU colleges absorb these costs at the college level (!)</p>
<p>Educational usage is entirely free, like for the <code>PAS2250</code> project you have been added to for this lecture.</p></li>
</ul>
</div>
<p align="center">
<img src="img/osc_logo.png" width="560">
</p>
<hr>
</section>
<section id="supercomputer-overview" class="level3">
<h3 class="anchored" data-anchor-id="supercomputer-overview">Supercomputer overview</h3>
<p align="center">
<img src="img/cluster_login.jpg" width="80%">
</p>
<hr>
</section>
<section id="osc-ondemand" class="level3">
<h3 class="anchored" data-anchor-id="osc-ondemand">OSC OnDemand</h3>
<p>The OSC OnDemand web portal allows you to use a web browser to access OSC resources such as:</p>
<ul>
<li><p>A file browser where you can also create and rename folders and files, and download and upload files. <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p></li>
<li><p>A Unix shell</p></li>
<li><p>More than a dozen different “Interactive Apps”, or programs with a GUI, such as RStudio, Jupyter, QGIS, and more.</p></li>
</ul>
<hr>
</section>
<section id="connecting-to-osc-with-ondemand" class="level3">
<h3 class="anchored" data-anchor-id="connecting-to-osc-with-ondemand">Connecting to OSC with OnDemand</h3>
<ul>
<li>Go to <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>, and log in with your OSC credentials.</li>
</ul>
<p align="center">
<img src="img/ondemand_homepage.png" width="80%">
</p>
<hr>
</section>
<section id="ondemand-files-menu" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-files-menu">OnDemand “Files” menu</h3>
<p>Choose a folder as a starting point for the file browser:</p>
<p align="center">
<img src="img/ondemand_files.png" width="60%">
</p>
<hr>
</section>
<section id="ondemand-files-menu-1" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-files-menu-1">OnDemand “Files” menu</h3>
<p>Here you can view, create and rename folders and files, and download and upload files:</p>
<p align="center">
<img src="img/ondemand_files2.png" width="85%">
</p>
<hr>
</section>
<section id="ondemand-clusters-menu" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-clusters-menu">OnDemand “Clusters” menu</h3>
<p>Here you can access a Unix shell on either of the two clusters:</p>
<p><br></p>
<p align="center">
<img src="img/ondemand_clusters.png" width="40%">
</p>
<p><br></p>
<p><em>(Since the two clusters share the file systems, and they have fairly similar capabilities, it generally doesn’t matter which cluster you connect to).</em></p>
<hr>
</section>
<section id="ondemand-clusters-menu-shell" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-clusters-menu-shell">OnDemand “Clusters” menu: Shell</h3>
<p>When you click on one of the shell options, a new browser tab with a shell will open. There are some welcome messages, and some storage usage/quota info, and then you get a prompt to type commands:</p>
<p align="center">
<img src="img/ondemand_terminal.png" width="95%">
</p>
<hr>
</section>
<section id="ondemand-clusters-menu-system-status" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-clusters-menu-system-status">OnDemand “Clusters” menu: System Status</h3>
<p>If you click on “System status”, you’ll get an overview of the current usage of the two clusters:</p>
<p><br></p>
<p align="center">
<img src="img/ondemand_systemstatus.png" width="100%">
</p>
<hr>
</section>
<section id="ondemand-interactive-apps-menu" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-interactive-apps-menu">OnDemand “Interactive Apps” menu</h3>
<p>Here you can mostly access programs with a GUI that will run on a compute node. We’ll try <code>RStudio Server</code> now (and <code>Code Server</code> a little later).</p>
<p align="center">
<img src="img/ondemand_interactiveapps.png" width="27%">
</p>
<hr>
</section>
<section id="ondemand-interactive-apps-rstudio-form" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-interactive-apps-rstudio-form">OnDemand “Interactive Apps”: RStudio Form</h3>
<p>Fill out this form to start an RStudio session.<br>
<em>This will run on a compute node and is therefore charged:</em> <em>for that reason,</em> <em>it needs the OSC account number so as to bill the correct account.</em></p>
<p align="center">
<img src="img/ondemand_rstudioform.png" width="43%">
</p>
<hr>
</section>
<section id="ondemand-interactive-apps-rstudio" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-interactive-apps-rstudio">OnDemand “Interactive Apps”: RStudio</h3>
<p>Once the top bar of the box like the one shown below turns green and says “Running”, you can click “<strong>Connect to RStudio Server</strong>” way at the bottom:</p>
<p align="center">
<img src="img/ondemand_rstudiocard.png" width="60%">
</p>
<hr>
</section>
<section id="ondemand-interactive-apps-rstudio-1" class="level3">
<h3 class="anchored" data-anchor-id="ondemand-interactive-apps-rstudio-1">OnDemand “Interactive Apps”: RStudio</h3>
<p>Now, you’ll have RStudio running in your browser!<br>
It looks just like the desktop app version you may be familiar with:</p>
<p align="center">
<img src="img/ondemand_rstudio.png" width="59%">
</p>
<hr>
</section>
<section id="what-is-different-at-osc-on-a-supercomputer" class="level3">
<h3 class="anchored" data-anchor-id="what-is-different-at-osc-on-a-supercomputer">What is different at OSC / on a supercomputer?</h3>
<div class="incremental">
<ul class="incremental">
<li><p><strong>Software</strong><br>
Because you don’t have administrator rights, and because the system is shared by so many people, you can’t install and use software “the regular way”.</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> For system-wide installed software, <strong>load it</strong> with <code>module</code> commands.</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> If something is not installed, ask OSC or use <em>Conda</em> or containers.</p></li>
</ul>
<p><br></p>
<ul class="incremental">
<li><strong>Login versus compute nodes</strong> (node ≈ computer)<br>
“Login nodes”, the nodes you end up on after logging in, are not meant for heavy computing and you have to <em>request access to “compute nodes”</em> to run most analyses.</li>
</ul>
<p><br></p>
<ul class="incremental">
<li><strong>“Non-interactive” usage is common, using a job scheduler (SLURM)</strong><br>
You submit your scripts to the SLURM queue and monitor the resulting jobs.</li>
</ul>
</div>
<hr>
</section>
<section id="overview-2" class="level3">
<h3 class="anchored" data-anchor-id="overview-2">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p>Command-line genomics?</p></li>
<li><p>Ohio Supercomputer Center (OSC) overview</p></li>
<li><p><strong>Command-line software</strong></p></li>
<li><p>The VS Code editor and the whole game</p></li>
</ol>
<hr>
</section>
<section id="an-example-of-a-command-line-program" class="level3">
<h3 class="anchored" data-anchor-id="an-example-of-a-command-line-program">An example of a command-line program</h3>
<p>A useful example of a genomics tool with a CLI is <strong>FastQC, a program for quality control of FASTQ files.</strong></p>
<p>It is ubiquitous because nearly all high-throughput sequencing data comes in FASTQ files, and your first step is always to check the quality of the reads.</p>
<p>FastQC produces visualizations and assessments of aspects of your reads such as adapter content, and, as shown below, <strong>mean base quality along the read</strong>:</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><img src="img/fastqc_good.png" width="85%"></p>
</div><div class="column" style="width:50%;">
<p><img src="img/fastqc_bad.png" width="85%"></p>
</div>
</div>
<hr>
</section>
<section id="running-fastqc" class="level3">
<h3 class="anchored" data-anchor-id="running-fastqc">Running FastQC</h3>
<div class="incremental">
<ul class="incremental">
<li><p>To run FastQC, you use the command <code>fastqc</code>.</p>
<p>Command-line programs are typically run non-interactively, so we don’t fire up the program first, and tell it what to do later, like we would with a program with a GUI.</p>
<p>Instead, we at once issue a complete set of instructions for the program to do what we would like it to.</p></li>
</ul>
<p><br></p>
<ul class="incremental">
<li><p>For example, say we want to analyze one of the FASTQ files that I put in <code>/fs/scratch/PAS2250/data</code>, with default FastQC settings.<br>
A complete FastQC command would be:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ fastqc /fs/scratch/PAS2250/data/sample1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>So, it is simply <code>fastqc</code> followed by a space and the name of the file!</p></li>
</ul>
</div>
<hr>
</section>
<section id="loading-the-fastqc-module-at-osc" class="level3">
<h3 class="anchored" data-anchor-id="loading-the-fastqc-module-at-osc">Loading the FastQC “module” at OSC</h3>
<div class="incremental">
<ul class="incremental">
<li><p>FastQC is available to us at OSC<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>, but we first have to <strong>load it</strong>. Here is what happens when we try to run the program in a fresh shell session at OSC:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ fastqc /fs/scratch/PAS2250/data/sample1.fastq.gz</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -bash: fastqc: command not found</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>We can load the software as follows:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ module load fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Now, let’s try again:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ fastqc /fs/scratch/PAS2250/data/sample1.fastq.gz</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Started analysis of sample1.fastq.gz</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Approx 5% complete for sample1.fastq.gz</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Approx 10% complete for sample1.fastq.gz</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Approx 15% complete for sample1.fastq.gz</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [truncated]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Success!</p></li>
</ul>
</div>
<hr>
</section>
<section id="something-is-missing-here" class="level3">
<h3 class="anchored" data-anchor-id="something-is-missing-here">Something is missing here</h3>
<p>I mentioned earlier that one benefit of running programs at the command-line is reproducibility – but how do we save the commands that we run?</p>
<ul>
<li><p>We need to not just save them, but to keep a detailed digital notebook that will enable us to redo our analysis.</p></li>
<li><p>We also need to wrap these commands in little scripts, so that we can run programs non-interactively and in parallel.</p></li>
</ul>
<p>For all of this, we will need a good <strong>text editor</strong>.</p>
<hr>
</section>
<section id="overview-3" class="level3">
<h3 class="anchored" data-anchor-id="overview-3">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p>Command-line genomics?</p></li>
<li><p>Ohio Supercomputer Center (OSC) overview</p></li>
<li><p>Command-line software</p></li>
<li><p><strong>The VS Code editor and the whole game</strong></p></li>
</ol>
<hr>
</section>
<section id="the-vs-code-text-editor" class="level3">
<h3 class="anchored" data-anchor-id="the-vs-code-text-editor">The VS Code text editor</h3>
<p>VS Code (in full, “Visual Studio Code”) is a nice modern GUI-based text editor. <a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a><br>
We can use a version of this editor (often referred to as <em>Code Server</em>) in our browser through OSC OnDemand.</p>
<p>Because it also has an integrated terminal to access a Unix shell, this setup effectively combines the 3 aspects of command-line genomics:</p>
<ul>
<li>Supercomputer</li>
<li>Unix shell</li>
<li>Text editor</li>
</ul>
<p><br></p>
<p><em>(And while I personally prefer RStudio for R, you can also run that through</em> <em>VS Code)</em>.</p>
<hr>
</section>
<section id="starting-up-vs-code-in-ondemand" class="level3">
<h3 class="anchored" data-anchor-id="starting-up-vs-code-in-ondemand">Starting up VS Code in OnDemand</h3>
<ol type="1">
<li>Log in to OSC’s OnDemand portal at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>.</li>
<li>In the blue top bar, select <code>Interactive Apps</code> and then near the bottom of the dropdown menu, click <code>Code Server</code>.</li>
<li>In the form that appears on a new page:
<ul>
<li>Select project <code>PAS2250</code></li>
<li>No need to change “Number of hours” and “Working Directory”</li>
<li>Make sure the “Codeserver Version” is <code>4.8</code>.</li>
</ul></li>
<li>On the next page, once the top bar of the box has turned green and says <code>Runnning</code>, click <code>Connect to VS Code</code>.</li>
</ol>
<hr>
</section>
<section id="starting-up-vs-code-in-ondemand-1" class="level3">
<h3 class="anchored" data-anchor-id="starting-up-vs-code-in-ondemand-1">Starting up VS Code in OnDemand</h3>
<p align="center">
<img src="img/vscode_form.png" width="60%">
</p>
<hr>
</section>
<section id="starting-up-vs-code-in-ondemand-2" class="level3">
<h3 class="anchored" data-anchor-id="starting-up-vs-code-in-ondemand-2">Starting up VS Code in OnDemand</h3>
<p>Once the session is running, you can click “Connect to VS Code”:</p>
<p align="center">
<img src="img/vscode_card.png" width="80%">
</p>
<hr>
</section>
<section id="starting-up-vs-code-in-ondemand-3" class="level3">
<h3 class="anchored" data-anchor-id="starting-up-vs-code-in-ondemand-3">Starting up VS Code in OnDemand</h3>
<p>You’ll likely get these two pop-ups – click “Yes” and “Don’t Show Again”, respectively:</p>
<p align="center">
<img src="img/vscode-trust.png" width="45%">
</p>
<p align="center">
<img src="img/vscode-git.png" width="45%">
</p>
<hr>
</section>
<section id="vs-code-basics-side-bar-with-file-browser" class="level3">
<h3 class="anchored" data-anchor-id="vs-code-basics-side-bar-with-file-browser">VS Code Basics: Side bar with file browser</h3>
<p align="center">
<img src="img/vscode-welcome_ed.png" width="55%">
</p>
<hr>
</section>
<section id="vs-code-basics-editor-pane" class="level3">
<h3 class="anchored" data-anchor-id="vs-code-basics-editor-pane">VS Code Basics: Editor pane</h3>
<p>The main part of the VS Code is the <em>editor pane</em>.</p>
<p>Whenever you open VS Code, a editor pane tab with a <code>Get Started</code> document is automatically opened. We can use this document to open a new text file by clicking <code>New file</code> below <code>Start</code>, which opens as a second tab in the editor pane.</p>
<p align="center">
<img src="img/vscode_editor.png" width="80%">
</p>
<hr>
</section>
<section id="vs-code-basics-terminal" class="level3">
<h3 class="anchored" data-anchor-id="vs-code-basics-terminal">VS Code Basics: Terminal</h3>
<p>Open a terminal with a Unix shell by clicking &nbsp; <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; <code>Terminal</code> =&gt; <code>New Terminal</code>.<br>
<em>The great thing with this setup is that we can keep notes and write shell scripts</em> <em>in the same window as our shell and a file browser!</em></p>
<p align="center">
<img src="img/vscode_terminal.png" width="55%">
</p>
<hr>
</section>
<section id="the-whole-game-building-on-the-fastqc-example" class="level3">
<h3 class="anchored" data-anchor-id="the-whole-game-building-on-the-fastqc-example">The whole game: building on the FastQC example</h3>
<p>I’ve shown you the main pieces of the computational infrastructure for<br>
“<strong>command-line genomics</strong>”. We’ve seen a very basic example of loading and running a command-line tool at OSC.</p>
<p>The missing pieces for a fuller example of how such tools are run in the context of an actual genomics project are (if we stay with FastQC):</p>
<div class="incremental">
<ul class="incremental">
<li><p>Putting the command to run FastQC in a <strong>“shell script”</strong>.<br>
<em>(This used the same language (Bash) as the commands you’d type interactively,</em> <em>so at its most basic this involves pasting those commands into a text file.)</em></p></li>
<li><p><strong>Submitting the script</strong> to the SLURM scheduler queue as a “batch job”.<br>
<em>(At its most basic, this involves putting <code>sbatch</code> in front of the script name.)</em></p></li>
<li><p>To make use of the capabilities of the supercomputer and speeding up our analysis, we can <strong>submit multiple jobs <em>in parallel</em> using a loop</strong>.</p></li>
</ul>
</div>
<hr>
</section>
<section id="so-what-computational-skills-should-i-learn" class="level3">
<h3 class="anchored" data-anchor-id="so-what-computational-skills-should-i-learn">So what computational skills should I learn?</h3>
<div class="incremental">
<ul class="incremental">
<li><p><strong>The core skills</strong>:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Unix shell basics – the commonly used commands</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i>Some shell <em>scripting</em> basics</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> SLURM basics to submit and manage your batch jobs</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> R for “downstream”, statistical and visualization tasks</p></li>
<li><p><strong>When you start doing genomics projects more often</strong>:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Using <code>conda</code> or containers for software</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Unix data tools (<code>grep</code>, <code>sed</code>, <code>awk</code>, etc)</p></li>
<li><p><strong>When you want to become proficient in applied bioinformatics</strong>:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Version control with <code>git</code></p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> More advanced: formal workflow/pipeline management tools (e.g.&nbsp;<code>Nextflow</code>)</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> More advanced: Python (or advanced R) for custom data processing</p></li>
</ul>
</div>
<hr>
</section>
<section id="resources-for-further-learning-osc" class="level3">
<h3 class="anchored" data-anchor-id="resources-for-further-learning-osc">Resources for further learning: OSC</h3>
<ul>
<li>Tutorials and courses!
<ul>
<li><a href="https://khill42.github.io/OSC_IntroHPC/" class="uri">https://khill42.github.io/OSC_IntroHPC/</a></li>
<li><a href="https://osc.catalog.instructure.com/" class="uri">https://osc.catalog.instructure.com/</a></li>
</ul></li>
</ul>
<p><br></p>
<ul>
<li>OSC regularly has online introductory sessions, both overviews and more hands-on sessions – see the OSC Events page: <a href="https://www.osc.edu/events" class="uri">https://www.osc.edu/events</a>.<br>
They also have weekly office hours.</li>
</ul>
<p><br></p>
<ul>
<li>There is also some good introductory material at their <em>Getting Started</em> pages (<a href="https://www.osc.edu/resources/getting_started" class="uri">https://www.osc.edu/resources/getting_started</a>), as well as more specific <em>“HOWTO” pages</em> (<a href="https://www.osc.edu/resources/getting_started/howto" class="uri">https://www.osc.edu/resources/getting_started/howto</a>).</li>
</ul>
<hr>
</section>
<section id="resources-for-further-learning-osu" class="level3">
<h3 class="anchored" data-anchor-id="resources-for-further-learning-osu">Resources for further learning: OSU</h3>
<ul>
<li><p>OSU courses and workshops</p>
<ul>
<li><p>Jonathan Fresnedo Ramirez’s “Genome Analytics” course (<code>HCS 7004</code>)</p></li>
<li><p>Microbiome Informatics (<code>MICRBIO 8161</code>)</p></li>
<li><p>The online materials for the workshop “Command line basics for genomic analysis at OSC” that myself and Mike Sovic gave last August:<br>
<a href="https://mcic-osu.github.io/cl-workshop-22/" class="uri">https://mcic-osu.github.io/cl-workshop-22/</a></p></li>
<li><p>I have a course “Practical Computing Skills for Omics Data” (<code>PLNTPTH 5006</code>) that I am planning to teach in in Spring 2024. All materials for the 2021 version of this course (“Practical Computing Skills for Biologists”) are at: <a href="https://mcic-osu.github.io/pracs-sp21/" class="uri">https://mcic-osu.github.io/pracs-sp21/</a></p></li>
</ul></li>
</ul>
<hr>
</section>
<section id="resources-for-further-learning-osu-1" class="level3">
<h3 class="anchored" data-anchor-id="resources-for-further-learning-osu-1">Resources for further learning: OSU</h3>
<p>Some particularly useful books:</p>
<ul>
<li><p>The Linux Command Line (William Shotts, 2019)</p></li>
<li><p>Bioinformatics Data Skills (Vince Buffalo, 2015)</p></li>
<li><p>Computing Skills for Biologists: A Toolbox (Wilmes &amp; Allesino, 2019)</p></li>
<li><p>A Primer for Computational Biology (Shawn T. O’ Neil, 2019) <a href="https://open.oregonstate.education/computationalbiology/" class="uri">https://open.oregonstate.education/computationalbiology/</a></p></li>
</ul>
<hr>
</section>
<section id="mcic-bioinformatics-assistance" class="level3">
<h3 class="anchored" data-anchor-id="mcic-bioinformatics-assistance">MCIC bioinformatics assistance</h3>
<div class="incremental">
<ul class="incremental">
<li><p>I mainly provide <strong>research support</strong> — mostly to grad students and postdocs.</p>
<p>I work on many different types of projects, and by far the most common are <em>RNAseq</em> and <em>microbial metabarcoding</em>.</p></li>
<li><p>I typically <strong>help you do your analysis</strong> rather than do it for you. I also help with getting people up to speed with the above-mentioned skills.</p>
<p>In some cases, I do run analyses – for instance, with RNAseq and metabarcoding I can run a <strong>standardized, nearly automated workflow</strong> to get count data such that you can skip the “command-line genomics” part and work with R on your laptop.</p></li>
<li><p>What the MCIC gets in return is a <em>bioinformatics membership fee</em> and in case of significant contributions, <em>authorship</em>.</p></li>
<li><p>You can contact me at <a href="mailto:poelstra.1@osu.edu" class="email">poelstra.1@osu.edu</a> – though I will be away most of the rest of this month.</p></li>
</ul>
</div>
<hr>
</section>
<section id="any-questions" class="level3">
<h3 class="anchored" data-anchor-id="any-questions">Any questions?</h3>
<hr>
</section>
<section id="bonus-submitting-fastqc-batch-jobs" class="level3">
<h3 class="anchored" data-anchor-id="bonus-submitting-fastqc-batch-jobs">Bonus: Submitting FastQC batch jobs</h3>
<p>Due to time constraints, I have skipped over the details of these steps. But an example shell script to run FastQC can be found at <code>/fs/scratch/PAS2250/scripts/fastqc.sh</code>, which contains the following code:</p>
<div class="columns">
<div class="column" style="width:60%;">
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --account=PAS2250</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the software</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load fastqc</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Bash strict settings</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the placeholder variables</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="va">input_file</span><span class="op">=</span><span class="va">$1</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="va">output_dir</span><span class="op">=</span><span class="va">$2</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the output dir if needed</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="st">"</span><span class="va">$output_dir</span><span class="st">"</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Run FastQC</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span><span class="op">=</span><span class="st">"</span><span class="va">$output_dir</span><span class="st">"</span> <span class="st">"</span><span class="va">$input_file</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column" style="width:40%;">
<ul>
<li><p>Indicate it’s a Bash script and tell SLURM which OSC account to use</p></li>
<li><p>Strict settings make the script stop on failure</p></li>
<li><p>The script takes “arguments”, which are stored as placeholder variables <code>$1</code> and <code>$2</code>. This allows us to run the script for different files</p></li>
</ul>
</div>
</div>
<hr>
</section>
<section id="bonus-submitting-fastqc-batch-jobs-1" class="level3">
<h3 class="anchored" data-anchor-id="bonus-submitting-fastqc-batch-jobs-1">Bonus: Submitting FastQC batch jobs</h3>
<p>Here is how I would submit that script as a batch job to analyze one FASTQ file:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="va">fastq_file</span><span class="op">=</span>/fs/scratch/PAS2250/data/sample1.fastq.gz</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> /fs/scratch/PAS2250/scripts/fastqc.sh <span class="st">"</span><span class="va">$fastq_file</span><span class="st">"</span> results_jelmer</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323144</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br></p>
<p>And how you can loop over all FASTQ files to submit as many jobs in parallel as you have FASTQ files:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> fastq_file <span class="kw">in</span> /fs/scratch/PAS2250/data/<span class="pp">*</span>.fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">sbatch</span> /fs/scratch/PAS2250/scripts/fastqc.sh <span class="st">"</span><span class="va">$fastq_file</span><span class="st">"</span> results_jelmer</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323145</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323146</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323147</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323148</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323149</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323110</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323111</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Submitted batch job 16323112</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>And the most common Unix shell is the <strong>Bash shell</strong>, which runs the Bash language.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>But for large transfers, use a different interface: <em>Globus</em><a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>full list of installed software: <a href="https://www.osc.edu/resources/available_software/software_list" class="uri">https://www.osc.edu/resources/available_software/software_list</a><a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>While there are also command-line text editors, these make the learning curve for command-line genomics even steeper.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>